### ISSUE TITLE
How to rebuild Mayastor (docker images) with modifications to spdk, spdk-rs & io-engine components

### ISSUE THREAD

**Issue by imukherjee-sndk** [openebs/mayastor#1286]
**Describe the bug**
Hi!
I'm trying to rebuild Mayastor docker images after making changes to spdk, spdk-rs & io-engine components.
The modified codebases (Mayastor, spdk & spdk-rs) are hosted in our internal bitbucket git repos currently.

Changes made to Mayastor to obtain the modified code from internal bitbucket repos:
1. Modified .gitmodules to point to internal spdk-rs repo
2. Modifed nix/pkgs/libspdk/default.nix to point to the internal spdk repo (snippet below).
```
    src = fetchgit {
#      owner = "openebs";
#      repo = "spdk";
      url = "https://bitbucket.abcde.com/path/to/spdk_mayastor_csal.git";
      rev = "1aa1058b17b876c158669a751beefcdcc1fbdda0";
      sha256 = "sha256-0000000000000000000000000000000000000000000=";
      fetchSubmodules = true;
      branchName = "v22.05.x-mayastor-csal";
    };
```
Checking out code from internal repo requires username and password. I have enabled git-credentials but I still see a failure in fetching the code (unable to obtain username).

**To Reproduce**
Cmd used to rebuild the docker images:
$ nix-build -A images.mayastor-io-engine-dev

**Expected behavior**
Code for the above mentioned components should be fetched, build should succeed and docker images generated.

**Screenshots**
Current Failure:
[indro@fedora-black mayastor_csal]$ nix-build -A images.mayastor-io-engine-dev
building '/nix/store/jazracqkq34wy4kjkg60w05kjbns55py-io-engine-version.drv'...
these 6 derivations will be built:
  /nix/store/srdnyqb5sinf5ss83dw7xvwgi1vyd5lv-spdk_mayastor_csal-1aa1058.drv
  /nix/store/8p44jld8gxc55ihjd8iw8aq45c9z29x4-libspdk-dev-22.05-4c2049bea.drv
  /nix/store/0ygc89qzg55cpif1dd19v9m94svr9q01-io-engine.drv
  /nix/store/hd66knsriixhf6jx0y89aq31waabydps-docker-layer-mayastor-io-engine-dev.drv
  /nix/store/4g4sla7rj4kbzs6an35lh50akd3xqyyd-runtime-deps.drv
  /nix/store/s0p52dydhc9a4s4gv9j27hldcdldgh61-docker-image-mayastor-io-engine-dev.tar.gz.drv
building '/nix/store/srdnyqb5sinf5ss83dw7xvwgi1vyd5lv-spdk_mayastor_csal-1aa1058.drv'...
exporting https://bitbucket.abcde.com/path/to/spdk_mayastor_csal.git (rev 1aa1058b17b876c158669a751beefcdcc1fbdda0) into /nix/store/0bankwpppadiqng
i1v4hg4i2i9f72glp-spdk_mayastor_csal-1aa1058
Initialized empty Git repository in /nix/store/0bankwpppadiqngi1v4hg4i2i9f72glp-spdk_mayastor_csal-1aa1058/.git/
fatal: could not read Username for 'https://bitbucket.abcde.com': No such device or address
fatal: could not read Username for 'https://bitbucket.abcde.com': No such device or address
Unable to checkout 1aa1058b17b876c158669a751beefcdcc1fbdda0 from https://bitbucket.abcde.com/path/to/spdk_mayastor_csal.git.

** OS info (please complete the following information):**
 - Distro: Fedora 36
 - Kernel version 6.0.5-200.fc36.x86_64
 - MayaStor revision 95798f66a50627eaf9e5570cf790f80c08abc01e

**Additional context**
**Queries:**
1. Are the above listed changes to fetch code from internal repos correct? Or do I need to modifiy some additional .nix files or derivations?
2. I have already enabled & tested git-credentials on my system but as seen above, the build system is unable to get the username. What is the right way to do this?

**Comment by tiagolobocastro** [comment]
Have you tried using the ssh url instead?

**Comment by imukherjee-sndk** [comment]
@tiagolobocastro , I found a workaround to fetch the code.
1. Prefetch code to localhost (build machine)
2. Modify nix/pkgs/libspdk/default.nix to use a file:/// based URL
3. Added **"--option sandbox false"** to the original build cmdline (nix-build -A images.mayastor-io-engine-dev) to allow access to my pc's /home directory.

With these changes, I'm able to successfully fetch and build the modified SPDK code.

**Now I'm getting another error:**
building '/nix/store/vv391h824pvjipcdrd4fbsphsjfpgby7-io-engine.drv'...
unpacking sources
unpacking source archive /nix/store/wwza680p0l4qr3sip9fm465f7j2ns2x6-mayastor_csal
source root is mayastor_csal
Executing cargoSetupPostUnpackHook
unpacking source archive /nix/store/l94lkqc1bd0k8lbvc0rqiylvv0f36hkl-cargo-vendor-dir
Finished cargoSetupPostUnpackHook
patching sources
Executing cargoSetupPostPatchHook
Validating consistency between /tmp/nix-build-io-engine.drv-0/mayastor_csal//Cargo.lock and /tmp/nix-build-io-engine.drv-0/cargo-vendor-dir/Cargo.lock
Finished cargoSetupPostPatchHook
configuring
building
Executing cargoBuildHook
++ env CC_x86_64-unknown-linux-gnu=/nix/store/35pnk5kwi26m3ph2bc7dxwjnavpzl8cn-gcc-wrapper-10.3.0/bin/cc CXX_x86_64-unknown-linux-gnu=/nix/store/35pnk5kwi26m3ph2bc7dxwjnavpzl8cn-gcc-wrapper-10.3.0/bin/c++ CC_x86_64-unknown-linux-gnu=/nix/store/35pnk5kwi26m3ph2bc7dxwjnavpzl8cn-gcc-wrapper-10.3.0/bin/cc CXX_x86_64-unknown-linux-gnu=/nix/store/35pnk5kwi26m3ph2bc7dxwjnavpzl8cn-gcc-wrapper-10.3.0/bin/c++ cargo build -j 24 --target x86_64-unknown-linux-gnu --frozen --workspace --bins --exclude io-engine-bench
error: failed to load manifest for workspace member `/tmp/nix-build-io-engine.drv-0/mayastor_csal/io-engine`

Caused by:
  failed to load manifest for dependency `mayastor-api`

Caused by:
  failed to read `/tmp/nix-build-io-engine.drv-0/mayastor_csal/rpc/mayastor-api/Cargo.toml`

Caused by:
  No such file or directory (os error 2)
error: builder for '/nix/store/vv391h824pvjipcdrd4fbsphsjfpgby7-io-engine.drv' failed with exit code 101
error: 1 dependencies of derivation '/nix/store/49xqxkqf1n9da17zl1hl4n8b16xf4y6y-docker-layer-mayastor-io-engine-dev.drv' failed to build
error: 1 dependencies of derivation '/nix/store/rdap0ysy0aj3dlzhvfh1dv4i741x9iki-docker-image-mayastor-io-engine-dev.tar.gz.drv' failed to build

**Summarising the changes and build steps:**
1. git clone the modified mayastor code
2. Update the spdk default.nix
3. nix-build --option sandbox false -A images.mayastor-io-engine-dev


Any idea what I may be missing? Thanks!
